Rename std::ptr::Shared to NonNull and stabilize it

SimonSapin opened the pull request:
This implements the changes proposed at https://github.com/rust-lang/rust/issues/27730#issuecomment-352800629:

> * Rename `Shared<T>` to `NonNull<T>` and stabilize it. (Being in the `ptr` module is enough to say that it’s a pointer. I’m not very attached to this specific name though.)
> * Rename `Box<T>` methods ~~`from_unique`~~/`into_unique` to ~~`from_nonnull`~~/`into_nonnull` (or whatever names are deemed appropriate), replace `Unique<T>` with `NonNull<T>` in their signatures, and stabilize them.
> *  Replace `Unique<T>` with `NonNull<T>` in the signatures of methods of the `Alloc` trait.
> * Mark `Unique` “permanently-unstable” by replacing remaining occurrences of `#[unstable(feature = "unique", issue = "27730")]` with:
>
>   ```rust
>   #[unstable(feature = "ptr_internals", issue = "0", reason = "\
>       use NonNull instead and consider PhantomData<T> (if you also use #[may_dangle]), \
>       Send, and/or Sync")]
>   ```
>
>   (Maybe the `reason` string is only useful on the struct definition.) Ideally it would be made private to some crate instead, but it needs to be used in both liballoc and libstd.
> * (Leave `NonZero` and `Zeroable` unstable for now, and subject to future bikeshedding.)

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @KodrAus (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

SimonSapin commented:
That proposal so far has got a couple thumbs up and no negative comment or reaction. Maybe @rust-lang/libs can discuss and do a FCP in this thread?

leodasvacas commented:
Bikeshed: `nonnull` is an eyesore.

SimonSapin commented:
@leodasvacas What’s your preferred shed color?

By your use of lower case I assume you mean in the `Box` methods. I’m very open to renaming them, do you have a suggestions? Ideally these might replace `from_raw` and `into_raw`, but those are already stable with `*mut T`.

leodasvacas commented:
Sorry for being unclear, I just prefer `non_null` over `nonnull` in method names that's all.

SimonSapin commented:
Ok. So the methods would be `Box::from_non_null_raw(…)` and `Box::into_non_null_raw(…)`. I don’t have a strong opinion either way, let’s see what other people think.

scottmcm commented:
🚲🏡: since it's NonNull (not Nonnull), non_null makes sense over nonnull.

SimonSapin commented:
I’ve pushed a commit to rename `*_nonnull_raw` to `*_non_null_raw`. Note however that there is precedent in the name of the `core::nonzero` module.

ExpHP commented:
But `core::nonzero` is not stable, is it?

That said, riding in the :bike: lane, `non_zero` certainly seems awkward.  If I may suggest:

* `NotNull` instead of `NonNull`.
  * `Box::from_not_null`
* Embrace "nonzero" as a single word.

```rust
use ::std::ptr::NotNull;
use ::core::nonzero::Nonzero;

fn foo() {
    let raw = NotNull::new(&0).unwrap();
    let _ = Box::from_not_null_raw(raw);
}
```

SimonSapin commented:
Besides naming, this is also a good time to reconsider whether `Box::from_non_null_raw` and `Box::into_non_null_raw` should exist in the first place.

I added `into_non_null_raw` in order to remove a few `unsafe` blocks in `rc`, `arc`, and `linked_list`. They use `Box` to allocate but then only keep a `NonNull<T>` pointer. Without this function, they (and other users doing similar things) would need unsafe to assert that the result of `Box::into_raw()` is not null.

So I think there's value in keeping this conversion in  some form. It could be an impl of the `From` trait, but we generally avoid methods (as opposed to associated functions) on generic smart pointers as these could shadow methods of the contained type.

`Box::from_non_zero_raw` however I think we can remove. I added it at the same for symmetry, but it's not used at all in the tree. It can be replaced with `Box::from_raw(some_non_null.get_mut())`.

kennytm commented on src/libcore/ptr.rs:
```diff
@@ -2331,8 +2331,9 @@ impl<T: ?Sized> PartialOrd for *mut T {
 /// Unlike `*mut T`, `Unique<T>` is covariant over `T`. This should always be correct
 /// for any type which upholds Unique's aliasing requirements.
 #[allow(missing_debug_implementations)]
```
Could you impl Debug for NonNull (superseding #46792)?

goodmanjonathan commented on src/libcore/ptr.rs:
```diff
@@ -2530,55 +2543,49 @@ impl<T: ?Sized> Shared<T> {
 
-#[unstable(feature = "shared", issue = "27730")]
-impl<T: ?Sized> From<Unique<T>> for Shared<T> {
+#[stable(feature = "nonnull", since = "1.24.0")]
```
Shouldn't this be #[unstable(feature = "unique")]?

SimonSapin replied:
Maybe. Even if this `impl` is usable, it can’t be used if you can’t get a `Unique<T>` value in the first place on stable Rust. And anyway, IIRC stability attributes don’t work on `impl` blocks at the moment. (They’re insta-stable if the relevant traits and types are stable.)

goodmanjonathan replied:
Right, I was just afraid it might have other effects (e.g. like showing up in rustdoc). If not, I'm fine either way.

SimonSapin replied:
Unstable impls are visible in rustdoc too: https://doc.rust-lang.org/std/ptr/struct.Shared.html#impl-From%3CUnique%3CT%3E%3E

goodmanjonathan replied:
Alright, then.

SimonSapin commented:
* Rebased on top of #46792 
* Removed `Box::from_non_null_raw` (can be replaced with `Box::from_raw(x.get_mut())`)
* Renamed `Box::into_non_null_raw` to `Box::into_raw_non_null`

SimonSapin commented:
> It could be an impl of the From trait, but we generally avoid methods (as opposed to associated functions) on generic smart pointers as these could shadow methods of the contained type.

However we already have `impl<'a, T: ?Sized> From<&'a T> for NonNull<T>` and `impl<'a, T: ?Sized> From<&'a mut T> for NonNull<T>` which (indirectly) add an `.into()` method on `&T` and `&mut T`. So maybe it’s not such a problem in practice?

SimonSapin commented:
Pushed another commit (that we can remove from the PR if it turns out problematic) that replaces `Box::into_raw_non_null` with an impl of the `Into` trait. (Not `From` because coherence.)

kennytm commented:
I don't think it's a good idea to reintroduce an `Into` impl, esp. when the corresponding `From` can't exist...

SimonSapin commented:
@kennytm What about the `From<&T>` and `From<&mut T>` impls?

kennytm commented:
@SimonSapin Those sound fine to me. My concern is just about the `Into` impl. IIRC there was a PR that changed all `Into` impls to `From` impls, introducing a new `Into` impl is kinda regression in progress 😄.

SimonSapin commented:
Well, the reason both traits exist is because conversions between types from different crates can only be implemented with one or the other. I don’t see how adding a new `Into` impl is bad just because we removed (replaced) some other ones.

That said I’m not particularly attached to this one impl. It only avoid picking an awkward name for the `Box` associated function.

SimonSapin commented:
I removed that commit since it caused a warning (an error, with `#[deny(warnings)]`) for https://github.com/rust-lang/rust/issues/46205.

(Though this is not very informative :/

> TBD: write understandable version.

)

bors commented:
:umbrella: The latest upstream changes (presumably #47076) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
There is now a merge conflict at the nomicon submodule. This is gonna be a pain to deal with, so I’ll wait until some form of this PR is ready to be r+’ed to do that.

KodrAus commented:
Sorry, I'm just trying to catch up on the background behind this.

Is the idea that the only invariant `NonNull` expects to hold is that the pointer is not null, and it doesn't make any claims about shared or exclusive ownership of that pointer like `Unique` and `Shared` did?

SimonSapin commented:
TL;DR: it’s not super clear what shared vs exclusive ownership really meant anyway.

The names of `Unique<T>` and `Shared<T>` make sense in opposition to each other. With indeed the former being documented as “owning” a `T` value:

> It also implies the kind of strong aliasing guarantees an instance of `T` can expect: the referent of the pointer should not be modified without a unique path to its owning `Unique`.

rustc could in theory take advantage of this to let LLVM’s optimizer do more optimization, but my understanding is that it does not today because we’re not confident enough about what these guarantees mean exactly for `unsafe` code. Last I spoke with Gankro about this, he said this likely won’t happen until we have a formal memory model for Rust, and therefore `Unique` should remain an implementation detail of the standard library and not be stabilized, at least for now.

This leaves only `Shared<T>` up for stabilization. By itself, its current named doesn’t make as much sense. So in #27730 we discussed renaming it to `NonNullPtr` or `NonNull`, after its main characteristic. (Gankro is in favor of `Pointer` or `Ptr` but IMO that’s confusing, with `*const T` and `*mut T` also being called pointers.)

KodrAus commented on src/libcore/ptr.rs:
```diff
@@ -2521,6 +2545,7 @@ impl<T: ?Sized> Shared<T> {
     /// The resulting lifetime is bound to self so this behaves "as if"
     /// it were actually an instance of T that is getting borrowed. If a longer
     /// (unbound) lifetime is needed, use `&*my_ptr.ptr()`.
```
Should this be `&*my_ptr.as_ptr()`?

SimonSapin replied:
Good catch! Fixed. (This was a pre-existing bug.)

KodrAus commented on src/libcore/ptr.rs:
```diff
@@ -2530,55 +2555,49 @@ impl<T: ?Sized> Shared<T> {
     /// The resulting lifetime is bound to self so this behaves "as if"
     /// it were actually an instance of T that is getting borrowed. If a longer
     /// (unbound) lifetime is needed, use `&mut *my_ptr.ptr_mut()`.
```
Should this be `&*my_ptr.as_ptr()`?

SimonSapin replied:
Fixed as well.

KodrAus commented:
For reference, this is effectively the public API that's proposed:

```rust
pub struct NonNull<T: ?Sized> { ... }

impl<T: ?Sized> !Send for NonNull<T> { }
impl<T: ?Sized> !Sync for NonNull<T> { }

impl<'a, T: ?Sized> From<&'a mut T> for NonNull<T> { ... }
impl<'a, T: ?Sized> From<&'a T> for NonNull<T> { ... }

impl<T: Sized> NonNull<T> {
    pub fn empty() -> Self;
}

impl<T: ?Sized> NonNull<T> {
    pub const unsafe fn new_unchecked(ptr: *mut T) -> Self;

    pub fn new(ptr: *mut T) -> Option<Self>;

    pub fn as_ptr(self) -> *mut T;

    pub unsafe fn as_ref(&self) -> &T;

    pub unsafe fn as_mut(&mut self) -> &mut T;
}

impl<T: ?Sized> Box<T> {
    fn into_raw_non_null(b: Box<T>) -> NonNull<T>;
}

impl<T: ?Sized> Debug for NonNull<T> { ... }
impl<T: ?Sized> Pointer for NonNull<T> { ... }

impl<T: ?Sized> Clone for NonNull<T> { ... }
impl<T: ?Sized> Copy for NonNull<T> { }

impl<T: ?Sized, U: ?Sized> CoerceUnsized<NonNull<U>> for NonNull<T> where T: Unsize<U> { }
impl<T: RefUnwindSafe + ?Sized> UnwindSafe for NonNull<T> { }
```

aturon commented:
Nominating for libs team discussion during the meeting tomorrow.

sfackler commented on src/libcore/ptr.rs:
```diff
@@ -2406,112 +2412,130 @@ impl<T: ?Sized> Unique<T> {
+#[stable(feature = "nonnull", since = "1.24.0")]
+impl<T: ?Sized> fmt::Debug for NonNull<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:p}", self.as_ptr())
```
This should probably just call `fmt::Pointer::fmt(self, f)` to pass along formatting options.

SimonSapin replied:
Done. (I assume you meant `self.as_ptr()`.)

aturon commented:
The libs team discussed this today, and are in favor of moving forward with this type, modulo some small API nits around `empty`. Kicking off the process:

@fcpbot rfc merge

aturon commented:
@rfcbot fcp merge

aturon commented:
re: the `empty` API, people were a bit uneasy about the idea of `NonNull::empty`, which seems like an oxymoron. Are there use-cases outside of `Vec`? Is there a more clear name, or other way we can approach this?

Otherwise, though, the libs team is feeling good about the unification of `Shared`/`Unique` and overall direction of the API here.

kennytm commented:
Maybe rename `empty` to `dangling` from #45527?

sfackler commented:
I like `dangling`.

alexcrichton commented on src/liballoc/boxed.rs:
```diff
@@ -326,40 +295,40 @@ impl<T: ?Sized> Box<T> {
-               issue = "27730")]
+    #[stable(feature = "nonnull", since = "1.24.0")]
+    #[inline]
+    pub fn into_raw_non_null(b: Box<T>) -> NonNull<T> {
```
Could this perhaps be unstable for the first pass? Or do we think that it'll clearly get enough use to warrant stabilization in the first pass?

alexcrichton commented:
I'm also a fan of `dangling`! And other than that looks great to me!

SimonSapin commented:
* Changed `Debug` impls to preserve formatting options
* Reverted `Box::into_raw_non_null` to unstable
* Renamed `NonNull::empty` to `dangling`. I looked at renaming `Unique::empty` (which is unstable) too, but that requires coordinating changes in the nomicon and miri submodules, so I left a FIXME comment instead.

SimonSapin commented:
Made `Box::into_raw_non_null` have its own tracking issue: https://github.com/rust-lang/rust/issues/47336.

alexcrichton commented:
@SimonSapin want to rebase and I can r+?

SimonSapin commented:
I was gonna wait until the end of FCP, but that works for me. Because the conflict is in a submodule, https://github.com/rust-lang-nursery/nomicon/pull/55 needs to be merged first. (Note that that PR goes *into* a branch that’s not `master`)

alexcrichton commented:
@SimonSapin oh for PRs we don't actually wait for the end of FCP, we mostly just use it to ensure everyone's signed off. Once that happens though it's free to be r+'d!

@bors: r+

bors commented:
:pushpin: Commit 9c96bb0 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 9c96bb0d12ecb5046e4f2b0d311fb5c3390b1c86 -->

kennytm commented:
@bors r-

Travis failed. Missing `#![feature(box_into_raw_non_null)]` in liballoc.

<details>

```
[01:11:24] error[E0658]: use of unstable library feature 'box_into_raw_non_null' (see issue #47336)
[01:11:24]    --> liballoc/arc.rs:289:20
[01:11:24]     |
[01:11:24] 289 |         Arc { ptr: Box::into_raw_non_null(x), phantom: PhantomData }
[01:11:24]     |                    ^^^^^^^^^^^^^^^^^^^^^^
[01:11:24]     |
[01:11:24]     = help: add #![feature(box_into_raw_non_null)] to the crate attributes to enable
[01:11:24] 
[01:11:24] error[E0658]: use of unstable library feature 'box_into_raw_non_null' (see issue #47336)
[01:11:24]    --> liballoc/arc.rs:994:22
[01:11:24]     |
[01:11:24] 994 |                 ptr: Box::into_raw_non_null(box ArcInner {
[01:11:24]     |                      ^^^^^^^^^^^^^^^^^^^^^^
[01:11:24]     |
[01:11:24]     = help: add #![feature(box_into_raw_non_null)] to the crate attributes to enable
[01:11:24] 
[01:11:24] error[E0658]: use of unstable library feature 'box_into_raw_non_null' (see issue #47336)
[01:11:24]    --> liballoc/rc.rs:314:18
[01:11:24]     |
[01:11:24] 314 |             ptr: Box::into_raw_non_null(box RcBox {
[01:11:24]     |                  ^^^^^^^^^^^^^^^^^^^^^^
[01:11:24]     |
[01:11:24]     = help: add #![feature(box_into_raw_non_null)] to the crate attributes to enable
[01:11:24] 
[01:11:24] error[E0658]: use of unstable library feature 'box_into_raw_non_null' (see issue #47336)
[01:11:24]     --> liballoc/rc.rs:1193:22
[01:11:24]      |
[01:11:24] 1193 |                 ptr: Box::into_raw_non_null(box RcBox {
[01:11:24]      |                      ^^^^^^^^^^^^^^^^^^^^^^
[01:11:24]      |
[01:11:24]      = help: add #![feature(box_into_raw_non_null)] to the crate attributes to enable
[01:11:24] 
[01:11:25] error[E0658]: use of unstable library feature 'box_into_raw_non_null' (see issue #47336)
[01:11:25]    --> liballoc/linked_list.rs:160:29
[01:11:25]     |
[01:11:25] 160 |             let node = Some(Box::into_raw_non_null(node));
[01:11:25]     |                             ^^^^^^^^^^^^^^^^^^^^^^
[01:11:25]     |
[01:11:25]     = help: add #![feature(box_into_raw_non_null)] to the crate attributes to enable
[01:11:25] 
[01:11:25] error[E0658]: use of unstable library feature 'box_into_raw_non_null' (see issue #47336)
[01:11:25]    --> liballoc/linked_list.rs:195:29
[01:11:25]     |
[01:11:25] 195 |             let node = Some(Box::into_raw_non_null(node));
[01:11:25]     |                             ^^^^^^^^^^^^^^^^^^^^^^
[01:11:25]     |
[01:11:25]     = help: add #![feature(box_into_raw_non_null)] to the crate attributes to enable
[01:11:25] 
[01:11:25] error[E0658]: use of unstable library feature 'box_into_raw_non_null' (see issue #47336)
[01:11:25]    --> liballoc/linked_list.rs:989:33
[01:11:25]     |
[01:11:25] 989 |                 let node = Some(Box::into_raw_non_null(box Node {
[01:11:25]     |                                 ^^^^^^^^^^^^^^^^^^^^^^
[01:11:25]     |
[01:11:25]     = help: add #![feature(box_into_raw_non_null)] to the crate attributes to enable
[01:11:25] 
[01:11:28] error: aborting due to 7 previous errors
[01:11:28] 
[01:11:28] error: Could not compile `alloc`.
```

</details>

SimonSapin commented:
Should be fixed, though I don’t understand why the "normal" build of liballoc passed and only unit testing failed…

bors commented:
:pushpin: Commit 2cf5c6a has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 2cf5c6adaa318b80a4faec14d165f2ea031f60d1 -->

bors commented:
:hourglass: Testing commit 2cf5c6adaa318b80a4faec14d165f2ea031f60d1 with merge 46c78064e2826983f253876ebf87105bec0dc832...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6082)

SimonSapin commented:
```
Checking "basic.js" ... OK
Checking "enum-option.js" ... OK
Checking "fn-forget.js" ... OK
Checking "from_u.js" ... FAILED
==> Result not found in 'others': '{"path":"std::boxed::Box","name":"from_unique"}'
Checking "macro-print.js" ... OK
Checking "string-from_ut.js" ... OK
Checking "struct-vec.js" ... OK
command did not execute successfully: "C:\\Program Files (x86)\\nodejs\\node" "src/tools/rustdoc-js/tester.js" "x86_64-pc-windows-msvc"
expected success, got: exit code: 1
```

*Sigh*. This PR changes the expected result of an unrelated test (rustdoc seach results) that was added after the PR was created. Should be fixed now:

```diff
diff --git a/src/test/rustdoc-js/from_u.js b/src/test/rustdoc-js/from_u.js
index 920620a9ae..0296788f7a 100644
--- a/src/test/rustdoc-js/from_u.js
+++ b/src/test/rustdoc-js/from_u.js
@@ -15,7 +15,6 @@ const EXPECTED = {
         { 'path': 'std::char', 'name': 'from_u32' },
         { 'path': 'std::str', 'name': 'from_utf8' },
         { 'path': 'std::string::String', 'name': 'from_utf8' },
-        { 'path': 'std::boxed::Box', 'name': 'from_unique' },
         { 'path': 'std::i32', 'name': 'from_unsigned' },
         { 'path': 'std::i128', 'name': 'from_unsigned' },
     ],
```

eddyb commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit 602a445 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 602a445b92b37ec6af4d3d7f331e1a0d1360b8d2 -->

bors commented:
:hourglass: Testing commit 602a445b92b37ec6af4d3d7f331e1a0d1360b8d2 with merge bdda8d61151a91fcc95b059918dd834c8e7ac09e...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6085), [status-travis](https://travis-ci.org/rust-lang/rust/builds/331175854?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing bdda8d61151a91fcc95b059918dd834c8e7ac09e to master...

Gankro commented:
\o/ you did it!

joshlf commented:
Stop me if I should bring this up in a new issue, but how do folks feel about adding a conversion from `NonNull<T>` to `NonNull<U>`? It wouldn't add any extra unsafety because you can already do `NonNull::new(x.as_ptr() as *mut U).unwrap()`, and would basically be a cleaner way of writing that.

SimonSapin commented:
@joshlf It should probably be a new issue (or a PR ;)) but I’ve been thinking the same when considering whether every `*mut u8` in return types in the `Alloc` trait should be `NonNull<u8>` instead. So for though I think the need is less pressing as long as there are few APIs that return `NonNull<_>`: if you create it from a raw pointer anyway, you can cast with `as` beforehand.

joshlf commented:
Would've submitted a PR, but I see you've beaten me to it :)

On Jan 20, 2018 11:34 PM, "Simon Sapin" <notifications@github.com> wrote:

> @joshlf <https://github.com/joshlf> It should probably be a new issue (or
> a PR ;)) but I’ve been thinking the same when considering whether every *mut
> u8 in return types in the Alloc trait should be NonNull<u8> instead. So
> for though I think the need is less pressing as long as there are few APIs
> that return NonNull<_>: if you create it from a raw pointer anyway, you
> can cast with as beforehand.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/46952#issuecomment-359229736>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_2L_mZdcvSPvPHjYXaz1V6BSo1rguWks5tMuiagaJpZM4RLVAQ>
> .
>

SimonSapin commented:
Oops, I forgot to comment here about it. Yeah, since I was making other changes… :)
